File: rslib/src/sync/collection/progress.rs
// Copyright: Ankitects Pty Ltd and contributors
// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

use crate::error;
use crate::sync::collection::protocol::EmptyInput;
use crate::sync::collection::protocol::SyncProtocol;
use crate::sync::http_client::HttpSyncClient;
use crate::sync::login::SyncAuth;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SyncStage {
    Connecting,
    Syncing,
    Finalizing,
}

impl Default for SyncStage {
    fn default() -> Self {
        SyncStage::Connecting
    }
}

#[derive(Debug, Default, Clone, Copy)]
pub struct FullSyncProgress {
    pub transferred_bytes: usize,
    pub total_bytes: usize,
}

pub async fn sync_abort(auth: SyncAuth) -> error::Result<()> {
    HttpSyncClient::new(auth)
        .abort(EmptyInput::request())
        .await?
        .json()
}

pub type FullSyncProgressFn = Box<dyn FnMut(FullSyncProgress, bool) + Send + Sync + 'static>;

File: web/ts/lib/tslib/progress.ts
// Copyright: Ankitects Pty Ltd and contributors
// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

import type { Progress } from "@generated/anki/collection_pb";
import { latestProgress } from "@generated/backend";

export async function runWithBackendProgress<T>(
    callback: () => Promise<T>,
    onUpdate: (progress: Progress) => void,
): Promise<T> {
    let done = false;
    async function progressCallback() {
        const progress = await latestProgress({});
        onUpdate(progress);
        if (done) {
            return;
        }
        setTimeout(progressCallback, 100);
    }
    setTimeout(progressCallback, 100);
    try {
        return await callback();
    } finally {
        done = true;
    }
}
